gem/optimise.py

```diff
@@ -229,34 +255,101 @@ def delta_elimination(sum_indices, factors):
     return sum_indices, factors
 
 
-def associate_product(factors):
-    """Apply associativity rules to construct an operation-minimal product tree.
+ExpressionAndFlopCount = namedtuple('ExpressionAndFlopCount', ['expression', 'flops'])
+
 
-    For best performance give factors that have different set of free indices.
+def associate_operation(operands, type):
+    """Apply associativity rules to associative operations (e.g. summation and
+    product) to construct an operation-minimal operation tree.
+
+    :arg operands: list of operands
+    :arg type: GEM class which is associative, e.g. Sum, Product
+
+    :returns: named tuple ExpressionAndFlopCount, with expression holds the
+              resultant expression, and flops holds its number of flops
     """
-    if len(factors) > 32:
+    # group operands by their free indices
+    if type == Sum:
+        base = Zero()
+    elif type == Product:
+        base = one
+    else:
+        raise AssertionError(
+            "Unknown associative operation type {0}!".format(type))
+    groups = OrderedDict()
+    for operand in operands:
+        key = frozenset(operand.free_indices)
+        groups.setdefault(key, []).append(operand)
+    operands = [reduce(type, _value, base) for _value in itervalues(groups)]
```

Comment:
Yes, you shall have a generic grouping function and a generic association function, rather than one function that does both and fails to be actually generic.